### ISSUE TITLE
cdk-virtual-scroll-viewport  - support window scroll

### ISSUE THREAD

**Issue by abeymg** [angular/components#13862]
#### Bug, feature request, or proposal:

Feature request

#### What is the expected behavior?
Allow the scroll of the cdk-virtual-scroll-viewport to be controlled by the top level window scrollbar.

See example at the link below. (Go to that page and click on parent scroll to see desired behavior)
https://rintoj.github.io/ngx-virtual-scroller

Another example of window scrolling in the context of virtual scrolling
https://bvaughn.github.io/react-virtualized/#/components/WindowScroller

#### What is the current behavior?
Specifying a parent scroll does not seem to be supported, from what I can tell.

#### What is the use-case or motivation for changing an existing behavior?
Having a single top level window/page scroll is standard in a lot of SPAs. The virtual-scroll would be a lot more useful in these cases, if it could use the same scroll bar, otherwise often times you end up with two scrollbars (one for the virt-scroll and the other for the page, or else a smaller inset container for the virt-scroll with a scrollbar), which is less than ideal.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular: 7.0.1
Material: 7.0.2

**Comment by LeParadoxHD** [comment]
That would be very useful!

**Comment by vahidvdn** [comment]
+1

**Comment by skatestyle** [comment]
This should definitely be implemented soon!

**Comment by rmedaer** [comment]
@mmalerba I did a small POC to support both window and closest relative parent virtual scrolling.

Actually it's not yet ready for merge. It's still missing tests, examples, and even implementations... But at least it gives a proposal for the implementation. I would be really interested to get your feedback.

In my use-case it fits well. Hope you appreciate.

Kind regards,

**Comment by rmedaer** [comment]
Yeah ! I committed on [master branch in my fork](https://github.com/rmedaer/material2/commit/1d024d6ba0c85e3deb8e08c1a3635bd9570e396d). I know it's bad but again my plan was just to do a test and it became probably more ..

**Comment by mmalerba** [comment]
@rmedaer Overall I like the approach, though parts of the implementation I'm not too sure about. I also think we could make the nested viewport more flexible by either making a directive that can be used to mark the parent, or using an `@Input()` that lets the user specify the scrolling container (e.g. `viewportRoot=".some-ancestor-element"`), rather than just assuming it's the direct parent.

@jelbourn Do you have any thoughts about this proposal?

**Comment by jelbourn** [comment]
Could you summarize the approach?

**Comment by vahidvdn** [comment]
@hezzyg You can use [this ](https://github.com/rintoj/ngx-virtual-scroller)module. It already supports `window` scroll.

**Comment by rmedaer** [comment]
@hezzyg otherwise if @mmalerba and/or @jelbourn are OK with my approach, I can deliver a patch ASAP. But I'm still waiting their (no)go to do it (or not). ;-)

**Comment by hezzyg** [comment]
> @hezzyg otherwise if @mmalerba and/or @jelbourn are OK with my approach, I can deliver a patch ASAP. But I'm still waiting their (no)go to do it (or not). ;-)

Thanks @rmedaer . I do prefer a solution integrated into Angular Material for better support in the future. So I hope your suggestion will get there asap :-)

**Comment by hezzyg** [comment]
> @hezzyg You can use [this ](https://github.com/rintoj/ngx-virtual-scroller)module. It's already supported `window` scroll.

I'll have a look. thanks @vahidvdn

**Comment by mmalerba** [comment]
@rmedaer yes I'd like to do this, please send a PR and mention me on it, thanks!

**Comment by jeetparikh** [comment]
This would be a really useful feature. We are unable to use this at the moment because we have a scrollbar at a different level.

**Comment by levitatejay** [comment]
@rmedaer @mmalerba 

> @hezzyg otherwise if @mmalerba and/or @jelbourn are OK with my approach, I can deliver a patch ASAP. But I'm still waiting their (no)go to do it (or not). ;-)

> @rmedaer yes I'd like to do this, please send a PR and mention me on it, thanks!

Is there a pull request for the patch? 
I really need to implement a window scrolling CDK viewport so I'd be grateful for any help!

**Comment by mmalerba** [comment]
There is a PR, https://github.com/angular/material2/pull/14287, but it needs some changes and the author hasn't responded in quite a while

**Comment by jeetparikh** [comment]
@mmalerba the PR #14287 is simply introducing the virtual adapter but is there any PR to support scrolling at parent level or window level for `cdk-virtual-scroll-viewport`?

**Comment by mmalerba** [comment]
@jeetparikh yeah my bad, I saw that it was linked to this issue but didn't look closely enough, that PR is unrelated. Deleting my previous comment to avoid confusion for people reading this later.

I thought I remember seeing a PR related to this, but it may have just been a proof of concept on someone's branch

**Comment by jeetparikh** [comment]
@rmedaer thank you for sharing that. Does that work without any tweaks? Also, does it work with `cdk-table`? or I would have to make amendments?

**Comment by rmedaer** [comment]
Yes it should work out of the box by adding `window` to `cdk-virtual-scroll-viewport` component (see `cdk-virtual-scroll-viewport[window]` directive).

I don't know about `cdk-table`... You'll probably have to use `CdkVirtualScrollNestedViewport` directive to  deal with parent relative positioning... but again I don't know much more about that.

**Comment by jeetparikh** [comment]
thanks @rmedaer .. what exactly does the `nested viewport support`?

**Comment by rmedaer** [comment]
Instead of using existing viewport from `cdk-virtual-scroll-viewport`, it takes the [closest relative parent (`nativeElement.offsetParent`)](https://github.com/rmedaer/material2/commit/1d024d6ba0c85e3deb8e08c1a3635bd9570e396d#diff-be99b7f48cc2cf95c4751a473c408de6R59) as reference. Later [it uses `offsetTop` property](https://github.com/rmedaer/material2/commit/1d024d6ba0c85e3deb8e08c1a3635bd9570e396d#diff-be99b7f48cc2cf95c4751a473c408de6R87) to measure the scroll offset (c.f. `measureScrollOffset` method). Because my code is a POC it's currently implemented only to measure from "top".

**Comment by jeetparikh** [comment]
sounds good.. thanks @rmedaer I l have a look :)

**Comment by hanishmaram** [comment]
@rmedaer 

> @jeetparikh if it's really urgent, you can [take a look at this commit](https://github.com/rmedaer/material2/commit/1d024d6ba0c85e3deb8e08c1a3635bd9570e396d). It's the POC I made to test virtual scrolling on window.
> Btw I don't have so much time to do a PR for now but at least you can already play with it !

@rmedaer  How can I use this in my project?

**Comment by hanishmaram** [comment]
This is very feature. We are not able to use virtual scroll as scroll bars are at different levels.

**Comment by danghuynhdaty** [comment]
Will this feature be implemented soon?

**Comment by shanehsu** [comment]
@rmedaer Will you be open to fixing your previous #14287 ? It seems to have some [suggested changes here][0] I think the idea looks great!

[0]: https://github.com/angular/components/pull/14287/files

**Comment by ShakurOo** [comment]
One year later, there is no news about this feature :(

**Comment by Celtian** [comment]
> @rmedaer
> 
> > @jeetparikh if it's really urgent, you can [take a look at this commit](https://github.com/rmedaer/material2/commit/1d024d6ba0c85e3deb8e08c1a3635bd9570e396d). It's the POC I made to test virtual scrolling on window.
> > Btw I don't have so much time to do a PR for now but at least you can already play with it !
> 
> @rmedaer How can I use this in my project?


@hanishmaram : 

**Recipe for you:**

1. Build from source
`git clone https://github.com/rmedaer/material2
  && cd material2
  && npm run install && npm run deploy`

2. Then you need content into your repo, just this content
`material2/dist/packages (just cdk folder is enough)`

3. Installation into your project
`yarn add ${github_user}/${github_repo}`

4. Import into your project:
`import { ScrollingModule } from '${github_user}/${github_repo}/cdk/scrolling';`

@rmedaer 

Hi,

can you provide an example how to use window scroll? For my purpose, I can use "nested" but "window" is more optimal.


![Screenshot_20191230-220434](https://user-images.githubusercontent.com/18282249/71600728-99723c00-2b50-11ea-987c-15d56c2ac951.jpg)

The last small problem is, when you are not using window.scroll, adress bar on anrdoid cannot be hidden automatically when you scroll down.

I hope that this feature will be available soon.

Thanks, 

Celtian

**Comment by pjjroux** [comment]
What is the current status on this? Still being looked at?

**Comment by limes** [comment]
Guys, don't make me move to React :) Please write, is there a chance to do this task in 2020? Who needs to buy a beer? :D

**Comment by vahidvdn** [comment]
@limes If you are gonna change your framework just because of a small built-in feature, you will move for your entire life.
Just check [this module](https://github.com/rintoj/ngx-virtual-scroller) instead if you need this feature.

**Comment by limes** [comment]
@vahidvdn 
My adventure with Angular has been going on for several years and sometimes such small nuisance can drive me crazy. I have an application that is designed to be displayed as desktop / mobile / PWA / Electron / Android / iOS. One code base. The last thing I miss is hiding the address bar (while scrolling) in mobile browser. There is no other way but to scroll the window object. I tested the module you're writing about. The problem is that scrolling mode on the window does not work with https://material.angular.io/components/sidenav/overview. Angular Material is very important for me because of the great support for disabled people. That's why I would like to know if scrolling on window support will be available this year or I have to look for completely different solutions having in mind the parallel support of many different platforms.

**Comment by vahidvdn** [comment]
@limes Good point. But the important thing is that maybe this is a much bigger problem in reactjs. Because that's a library and even a lot of important things like `router` component is not coming out of the box (but made by the community). So even in react you will gonna use a 3rd party library for your virtual scroll. (You may have some problems or may not)

**Comment by limes** [comment]
@vahidvdn And that's the reason I use Angular. I like the fact that I have everything in the box and I don't have to look for more libraries and watch out for dependencies. Just this little problem with scroll on the window @ mobile :)

**Comment by iBaff** [comment]
> @vahidvdn
> [...] I tested the module you're writing about. The problem is that scrolling mode on the window does not work with https://material.angular.io/components/sidenav/overview. [...]

I implemented [ngx-virtual-scroller](https://github.com/rintoj/ngx-virtual-scroller) with a window scroll and [mat-sidenav](https://material.angular.io/components/sidenav/overview). It works perfectly now (URL Bar is hiding on iOS and Android), after getting the html and css set correctly (which was a pain).

But still I think supporting scroll on the document.window is essential for mobile for the reasons mentioned.

**Comment by limes** [comment]
> It works perfectly now (URL Bar is hiding on iOS and Android), after getting the html and css set correctly (which was a pain).

@iBaff Can you show me how you fixed it?

**Comment by dmc31a42** [comment]
@limes 
please check difference (div native element in sidenav-content) between
https://stackblitz.com/edit/angular-virtual-scroll-with-sidenav-work
and
https://stackblitz.com/edit/angular-virtual-scroll-with-sidenav-not-work

**Comment by dmc31a42** [comment]
@limes 
URL bar hiding work
https://stackblitz.com/edit/angular-virtual-scroll-with-sidenav-work-with-urlbar-hide

**Comment by limes** [comment]
@dmc31a42 

https://angular-virtual-scroll-with-sidenav-work-with-urlbar-hide.stackblitz.io/
https://angular-virtual-scroll-with-sidenav-work.stackblitz.io/

Maybe I'm doing something wrong, but no pages under the above addresses hide the address bar when I'm scrolling them. I tested on real iPhone/Chrome

**Comment by dmc31a42** [comment]
@limes I have checked. I'll tell you when find solution

**Comment by dmc31a42** [comment]
@limes I fixed https://angular-virtual-scroll-with-sidenav-work-with-urlbar-hide.stackblitz.io/ work in iOS
Key point is,
1. make 'overflow' of mat-sidenav-container and mat-sidenav-content to 'visible' to fix virtual-scroller does not work in sidenav
```css
mat-sidenav-container {
  overflow: visible;
}
mat-sidenav-content {
  overflow: visible;
}
```
2. use scroll.window to work url hide

**Comment by limes** [comment]
@dmc31a42 Thank you! 👍

**Comment by tomasdev** [comment]
@mmalerba is there any update on timeline / prioritization of this feature?

**Comment by eng-dibo** [comment]
I am searching for this approach. nice to see this issue.
is there any updates?

**Comment by jdussouillez** [comment]
Any updates on this ?

**Comment by Aymeriic** [comment]
An update would be really great

**Comment by sthan001** [comment]
What i feel is that, 
1. By default if we don't control overflow (overflow: auto) at cdk-virtual-scroll-spacer, that carries to parent element all the way to window.
Now that is missing in the cdk-virtual-scroll-viewport element. and cdk-virtual-scroll-spacer should accept the element be it window, of element in local component, or element from parent component as scrollable. A attribute possibly?
2. the cdk-virtual-scroll-viewport should be calculating header and footer size to occupy the cdk-virtual-scroll-spacer. Basically accept the header, footer and any other intermediate element height to be considered to have complete scroll option. Again via attribue.
3. Either spacer to take care of the complete page available to scroll or we can set height of cdk-virtual-scroll-content-wrapper excluding header/breadcrumb/footer including media query adjustments.
4. as much as possible the position of cdk-virtual-scroll-viewport should be not fixed, or absolute. because some application may not able to reuse at their component level depth if that is the case.

Angular doing lot, not sure why this feature is on hold for long time. And again, suggesting alternative 3rd party library is another overhead. These all angular can fix and sell itself well among industries.

**Comment by ursaciuc-adrian** [comment]
Until this is merged i found a quick workaround for this. You can create a new custom scrolling strategy which is basically the same as the fixed size scrolling strategy, but for the viewPort height you use the hight of the window and for the scroll offset you use the pageYOffset of the window. 
And what is left to do is to listen to the scroll event on the page and update the redered range when this happens.
Here is a basic stackblitz with it. It might not work for every case, but it's good enough for what i need it to do. Maybe it will help others too.
[
https://stackblitz.com/edit/window-virtual-scroll-strategy](url)

**Comment by ghenry22** [comment]
With Ionic Framework deprecating their own virtual-scroll in favour of CDK virtual-scroller the ability to scroll based on a parent element/window is essential in order for the CDK virtual-scroller to work with other common components like pull-to-refresh, infinite-scroll and iOS style auto expanding header styling in mobile apps.

Details from Ionic team:
https://github.com/ionic-team/ionic-docs/blob/liam-v6-docs/src/pages/angular/virtual-scroll.md

I would be really great if we could pass in a scroll element to the virtual-scroller to use in a similar way to how https://github.com/rintoj/ngx-virtual-scroller works.

**Comment by vahidvdn** [comment]
@ghenry22 ionic built-in virtual list component was literally useless. Hope it's going to work with the angular CDK virtual-scroller.

**Comment by ghenry22** [comment]
@vahidvdn I've used the Ionic virtual scroll component successfully in a number of cases.  It definitely has some quirks and never got the attention it needed but when you work around it's specific weirdness it does the job.

I agree though I am hoping that the cdk virtual scroller will provide a much more robust and more actively supported solution.  Only time will tell!

**Comment by syam38** [comment]
Any update on this feature?

**Comment by syam38** [comment]
> Until this is merged i found a quick workaround for this. You can create a new custom scrolling strategy which is basically the same as the fixed size scrolling strategy, but for the viewPort height you use the hight of the window and for the scroll offset you use the pageYOffset of the window. And what is left to do is to listen to the scroll event on the page and update the redered range when this happens. Here is a basic stackblitz with it. It might not work for every case, but it's good enough for what i need it to do. Maybe it will help others too. [ https://stackblitz.com/edit/window-virtual-scroll-strategy](url)

The stackblitz link is not working, could you please re share it @ursaciuc-adrian

**Comment by ursaciuc-adrian** [comment]
> The stackblitz link is not working, could you please re share it @ursaciuc-adrian
Updated the url

**Comment by vugar005** [comment]
+

**Comment by spike-rabbit** [comment]
@mmalerba Would you still accept a PR adding this feature? I would be happy to provide one following the suggestions by @rmedaer

**Comment by mmalerba** [comment]
@spike-rabbit Yes, I'm happy to look at a PR along the lines of what we discussed above.

**Comment by Harpush** [comment]
Will it allow to have two virtual scroll viewports under the same container and the scroll will be in the container? Thus allowing them both to render only what's shown relative to the shared parent container.

**Comment by spike-rabbit** [comment]
@mmalerba Great, thanks for the info. I plan to provide one around March. 

@Harpush Based on my current knowledge: yes. But I need to dive further into this to provide a reliable answer

**Comment by lincolnthree** [comment]
Thanks for all of you effort on this! Really looking forward to seeing this.

This feature is critical for a "native-like" feel in mobile apps with momentum scrolling. When scrolling locks up and you need to scroll again with a new gesture to continue past the virtual scroll viewport, or capture the gesture in the viewport instead of going past, it really interrupts the user experience.

I will be happy to test and contribute to this if there is a beta build to try!
